src/HealthChecks.AzureKeyVault/AzureKeyVaultHealthCheck.cs

```diff
@@ -11,71 +12,68 @@ namespace HealthChecks.AzureKeyVault
-                        await client.GetSecretAsync(_options.KeyVaultUrlBase, secret, cancellationToken);
-                    }
+                    var secretClient = CreateSecretClient();
+                    await secretClient.GetSecretAsync(_options.KeyVaultUrlBase, secret, cancellationToken);
+                }
 
-                    foreach (var key in _options.Keys)
-                    {
-                        await client.GetKeyAsync(_options.KeyVaultUrlBase, key, cancellationToken);
-                    }
+                foreach (var key in _options.Keys)
+                {
+                    var keyClient = CreateKeyClient();
+                    await keyClient.GetKeyAsync(_options.KeyVaultUrlBase, key, cancellationToken);
+                }
 
-                    foreach (var (key, checkExpired) in _options.Certificates)
+                foreach (var (key, checkExpired) in _options.Certificates)
+                {
+                    var certificateClient = CreateCertificateClient();
+                    var certificate = await certificateClient.GetCertificateAsync(key, cancellationToken);
+
+                    if (checkExpired && certificate.Value.Properties.ExpiresOn.HasValue)
                     {
-                        var certificate = await client.GetCertificateAsync(_options.KeyVaultUrlBase, key, cancellationToken);
+                        var expirationDate = certificate.Value.Properties.ExpiresOn.Value;
 
-                        if (checkExpired && certificate.Attributes.Expires.HasValue)
+                        if (expirationDate < DateTime.UtcNow)
                         {
-                            var expirationDate = certificate.Attributes.Expires.Value;
-
-                            if (expirationDate < DateTime.UtcNow)
-                            {
-                                throw new Exception($"The certificate with key {key} has expired with date {expirationDate}");
-                            }
+                            throw new Exception($"The certificate with key {key} has expired with date {expirationDate}");
                         }
                     }
                 }
+
                 return HealthCheckResult.Healthy();
             }
             catch (Exception ex)
             {
                 return new HealthCheckResult(context.Registration.FailureStatus, exception: ex);
             }
         }
-        private KeyVaultClient CreateClient()
+
+        private KeyClient CreateKeyClient()
         {
-            if (_options.UseManagedServiceIdentity)
-            {
-                var azureServiceTokenProvider = new AzureServiceTokenProvider();
-                return new KeyVaultClient(new KeyVaultClient.AuthenticationCallback(azureServiceTokenProvider.KeyVaultTokenCallback));
-            }
-            else
-            {
-                return new KeyVaultClient(GetToken);
-            }
+            return new KeyClient(_keyVaultUri, _azureCredential);

```

Comment:
As mentioned above, you should create these once.

Reply:
waiting for your response on my comment above